### ISSUE TITLE
Slow extraction

### ISSUE THREAD

**Issue by xiao-cheng** [lukehsiao/HardwareDeepDive#32]
TL;DR I think we should do most feature extractions before feed it to deepdive.

Calling my python table parsing script directly and save to multiple tsv files, doing number and word extraction this way should add no more than a few milliseconds for each file, rather than ~10 minutes for 100 cores as it is the case now. (I managed to run the current code with parallelized Greenplum). A correct setup should take less than 30 seconds even with millions of variables.

We have to keep in mind that python and distributed database have really high latency between stages and talking to each other - data has to be serialized and then parsed again via file or network. If data can be processed in a monolithic process then we should do that instead. DDlog is meant for simple filtering and big joins rather than long pipelines.

**Comment by xiao-cheng** [comment]
For comparison, I did process the entire Wikipedia by joining pages and redirect links with a Greenplum Deepdive setup - by just having ~3 stages. So I think the problem is not caused by my setup.

**Comment by xiao-cheng** [comment]
And Postgreql 9.6 should support parallel aggregation: http://blog.2ndquadrant.com/parallel-aggregate/

**Comment by lukehsiao** [comment]
You got the GreenPlum to work?

**Comment by xiao-cheng** [comment]
@lukehsiao Yes, I had to call a wrapper script created by Jaeho

**Comment by lukehsiao** [comment]
The other thing is that our pipeline is only a couple steps more than the spouse example right now. Yet, it has a big performance difference when doing features and supervision.

**Comment by xiao-cheng** [comment]
I think it's various joins and the UNNEST and ARRAY_AGG, these operations take ages in a database. You can just import my script directly and call parse.process(file,callback=your_extraction_function), where you do all the splitting and regex etc. that has no reason to involve a database and other files. Only take global statistics in DDlog

**Comment by xiao-cheng** [comment]
Just another idea - you move the extraction code into python (which I imagine shouldn't be that much work since you already have udf in python) so that the entire flow looks like this:

PDF ----raiders server----> tsv files of features -----I send you the files ran on all 1k pdfs of ~ 200MB ----> iterative dev on global features without deailing with PDFs/deep pipelines at all

**Comment by lukehsiao** [comment]
Fixed by (a6d9b19) where we did a big refactoring. We should decouple the table-extraction repo from this one, and just include the dump.tsv of the 100 pdf set, and then another .tsv for the larger sets.
